llvm/lib/Target/AMDGPU/GCNHazardRecognizer.cpp

```diff
@@ -441,42 +441,96 @@ using IsExpiredFn = function_ref<bool(const MachineInstr &, int WaitStates)>;
 using GetNumWaitStatesFn = function_ref<unsigned int(const MachineInstr &)>;
 
 // Search for a hazard in a block and its predecessors.
+// StateT must implement getHashValue() and isEqual().
 template <typename StateT>
 static bool
-hasHazard(StateT State,
+hasHazard(StateT InitialState,
           function_ref<HazardFnResult(StateT &, const MachineInstr &)> IsHazard,
           function_ref<void(StateT &, const MachineInstr &)> UpdateState,
-          const MachineBasicBlock *MBB,
-          MachineBasicBlock::const_reverse_instr_iterator I,
-          DenseSet<const MachineBasicBlock *> &Visited) {
-  for (auto E = MBB->instr_rend(); I != E; ++I) {
-    // No need to look at parent BUNDLE instructions.
-    if (I->isBundle())
-      continue;
-
-    switch (IsHazard(State, *I)) {
-    case HazardFound:
-      return true;
-    case HazardExpired:
-      return false;
-    default:
-      // Continue search
-      break;
+          const MachineBasicBlock *InitialMBB,
+          MachineBasicBlock::const_reverse_instr_iterator InitialI) {
+  SmallVector<std::pair<const MachineBasicBlock *, unsigned>> Worklist;
+  SmallDenseSet<std::pair<const MachineBasicBlock *, unsigned>> Visited;
+  SmallVector<std::pair<unsigned, unsigned>, 1> Collisions;
+  SmallDenseMap<unsigned, unsigned> StateHash2Idx;
+  SmallVector<StateT> States;
+
+  // States contains a vector of unique state structures.
```

Comment:
This feels overcomplicated and it's also similar to what `MapVector` does internally. I think you could probably reimplement this with a `MapVector<StateT, void>`, using `MapVector::getArrayRef` to convert between states and StateIdxs. This would have the (I claim) benefit that StateT would have to support standard DenseMap traits instead of the less standard `getHashValue` and `isEqual` methods.

Reply:
> however, the actual key type needs to be a pointer to the backing `SmallVector` _and_ an index into it, because traits methods are static -- if they were not static I could capture the `SmallVector` reference and keys would only need to be indexes into the array.

If you use `std::deque` then you can work with pointers to the `StateT` objects, not indexes, because `deque` promises not to move existing allocations as it grows.

Reply:
Would you accept the current implementation using `SmallVector` and `SmallDenseMap`?
To re-iterate, this version while not *the best* is still better than current implementation in both performance and correctness.